### PR TITLE
WIP: Adding abstract method to browsertab.py Closes #6121

### LINKED ISSUE #6121: Get rid of browser.inspector.create
The local imports in `browser.inspector.create` are problematic, e.g. because an old qutebrowser instance could import new files after an upgrade.

Instead, `AbstractTabPrivate` in `browsertab.py` should have an abstract `_init_inspector` method or so, which then is implemented in `WebKitTab` and `WebEngineTab`.

### PR DESCRIPTION
**Pull request by jbenge1**

### CODE REVIEW

File: qutebrowser/browser/browsertab.py [hunk 1]
```diff
@@ -806,6 +808,10 @@ def __init__(self, mode_manager: modeman.ModeManager,
         self._tab = tab
         self._mode_manager = mode_manager
 
+    @abstractmethod
```
Comment (The-Compiler) [inline qutebrowser/browser/browsertab.py:811]: From the [`@abstractmethod` docs](https://docs.python.org/3/library/abc.html#abc.abstractmethod):

> Using this decorator requires that the class’s metaclass is `ABCMeta` or is derived from it.

We don't use `abc` here, for two reasons:

- Custom metaclasses like this usually don't play well with PyQt classes and inheriting from them.
- We don't need the runtime validation. If we have a normal method raising `NotImplementedError`, the `pylint` linter we use already makes sure it does get overridden.

File: qutebrowser/browser/browsertab.py [hunk 2]
```diff
@@ -806,6 +808,10 @@ def __init__(self, mode_manager: modeman.ModeManager,
...
+    def __init_inspector(self):
```
Comment (The-Compiler) [inline qutebrowser/browser/browsertab.py:812]:
```suggestion
    def _init_inspector(self):
```

With the double underscore, you [turn on name mangling](https://docs.python.org/3/tutorial/classes.html#private-variables), making it impossible to override the class in subclasses.

File: qutebrowser/browser/browsertab.py [hunk 3]
```diff
@@ -806,6 +808,10 @@ def __init__(self, mode_manager: modeman.ModeManager,
...
+        pass
```
Comment (The-Compiler) [inline qutebrowser/browser/browsertab.py:813]:
```suggestion
        raise NotImplementedError
```

**Review by The-Compiler — CHANGES_REQUESTED** [review]
Thanks for getting the ball rolling! I added a couple of comments.

You will then need to:

- Override the abstract method in `WebKitTab` and `WebEngineTab`, based on the `create` function in `inspector.py`
- Adjust the `inspector.create` call in `toggle_inspector` to call the new `self._init_inspector` instead.

Finally, please move the method down to the other inspector code (`_recreate_inspector` and `toggle_inspector`, so that related methods are in the same place).

**Comment by jbenge1** [comment]
Thanks for the feedbaack, sorry if this issue seems a bit hand holdy. I'm still trying to get a handle on the code base (and figured taking on some easier issues was the best way)

**Review by The-Compiler — CHANGES_REQUESTED** [review]
Don't worry about the hand-holding - I like teaching Python to people (it's what I do as [part of my job](https://bruhin.software/)), and if I can do that for a qutebrowser contribution, that's even better!

A couple of open issues I can see:

- Methods need to have `self` as first argument, see e.g. [this article](https://medium.com/quick-code/understanding-self-in-python-a3704319e5f0)
- Before this change, we used `if objects.backend == ...` to decide between the two implementations in the old `create` function. There's actually a name for this kind of change: [replace conditional with polymorphism](https://refactoring.guru/replace-conditional-with-polymorphism). Since we now move the implementations into `WebKitTab` / `WebEngineTab`, we now automatically call the right variant of `_init_inspector`. Thus, the new method in `WebKitTab` should only contain the QtWebKit-specific `return ...` line, while the one in `WebEngineTab` should only contain the QtWebEngine-specific code.
- After changing that, you will also be able to move the imports to the top of the respective files, and remove the "Importing modules here ..." comment. Since we're only importing `webkittab.py` / `webenginetab.py` when we are running with the respective backend, those are allowed to have backend-specific imports at file-level.
- You can then also remove the `objects` import. The `miscwidgets` one you added was never needed (but I guess that was a copy-paste mistake).
- You can also get rid of the old `create` function in `inspector.py`, as it's not used anymore.
- Finally, note that there should be two blank lines between classes, and no spaces on those blank lines. You can see the respective warnings in the diff here on GitHub, but you should also be able to run `tox -e flake8` to see them locally.
- Don't worry about the `mypy` issues for now. I suspect they will go away with the `self` arguments added properly, but if not, let's look at those last. I'm not sure off-hand what the problem is there.

**Review by The-Compiler — CHANGES_REQUESTED** [review]
Thanks for the update! This is going into the right direction, but right now has various issues still - please look at the [linter annotations](https://github.com/qutebrowser/qutebrowser/pull/6199/files) for details, and feel free to ask if something is unclear.

**Comment by jbenge1** [comment]
Thanks for all the help! I did have a couple of questions. 
1. Maybe I'm misunderstanding what `*` means but I'm guessing I need to explicitly state `self` as my first argument? 
2. The linter suggests both `objects` and `miscwidgets` are undefined (as I removed the imports I'm assuming), specifically line 883 of `webenginetab.py` `miswidgets` is undefined but I don't see a reference to it anywhere in the file, am I misunderstanding the output from the linter annotations? Or have I missed the mark completely?

**Comment by The-Compiler** [comment]
> * Maybe I'm misunderstanding what `*` means but I'm guessing I need to explicitly state `self` as my first argument?

The `*` means that all following arguments should be [keyword-only](https://www.python.org/dev/peps/pep-3102/), i.e. that the function always needs to be called like ` _init_inspector(splitter=..., win_id=42, parent=...)` rather than `_init_inspector(..., 42, ...)` because it'd be unclear what those arguments are exactly (and it's easy to accidentally swap them).

Indeed the `self` argument should always be there explicitly, even if it's not used (and even if `*` is after it). Some projects like using [`@staticmethod`](https://docs.python.org/3/library/functions.html#staticmethod) in cases like this, but I don't really see the point of doing so.

So, long story short: Yep, add `self` as first argument.

> * The linter suggests both `objects` and `miscwidgets` are undefined (as I removed the imports I'm assuming), specifically line 883 of `webenginetab.py` `miswidgets` is undefined but I don't see a reference to it anywhere in the file, am I misunderstanding the output from the linter annotations? Or have I missed the mark completely?

I don't follow. The lines the linters complain about are using [`objects.debug_flags`](https://github.com/qutebrowser/qutebrowser/pull/6199/files#diff-06974ebd62dda0821470d71ac10a25e9d0be7b032a042d02b6c094c30b11ecfeR537) and [`miscwidgets.FullscreenNotification`](https://github.com/qutebrowser/qutebrowser/pull/6199/files#diff-06974ebd62dda0821470d71ac10a25e9d0be7b032a042d02b6c094c30b11ecfeR883), and those are both in `webenginetab.py` where you removed those imports.

**Comment by jbenge1** [comment]
Regarding the second question, you're right I'm just being dense this morning.

**Comment by The-Compiler** [comment]
Looks like there are still some linter issues. After those are fixed I think this should be ready. :)

**Comment by The-Compiler** [comment]
Looks great now - thanks for the contribution and the quick updates!

**Pull request merged**